### ISSUE TITLE
twitter bootstrap modal as component

### ISSUE THREAD

**Issue by PavelPolyakov** [vuejs/Discussion#247]
Could someone give me a hint, how can I create a component from the twitter bootstrap modal.

Assuming that his is the html of the modal:
```html
<div class="modal fade" id="addVerb">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
                <h4 class="modal-title">Modal title</h4>
            </div>
            <div class="modal-body">
                <p>One fine body&hellip;</p>
            </div>
            <div class="modal-footer">
                <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                <button type="button" class="btn btn-primary" v-on="click: addVerbSave">Save changes</button>
            </div>
        </div><!-- /.modal-content -->
    </div><!-- /.modal-dialog -->
</div><!-- /.modal -->
```

When I want to achieve:
1. When I click a link, the component is created dynamically. E.g. it's not present in the template of the main component before the modal is initialised.
2. Component can interact with the data of the main component - the one from which it was created.

What I can assume - I should describe a standalone component, like we do usually, let's call it "modal".
If I put the ```<modal prop={{@ parentData}}></modal>``` to the main component template - I would be able to use it.

But - how can I do that dynamically? If I want to put/remote the modal component on demand? Should I somehow manipulate with this.$el ??? Where can I read about it?

Any advice is appreciated.

Regards,

**Comment by PavelPolyakov** [comment]
If I do the next, inside the component method:
```javascript
$(this.$el).append('<foobar></foobar>');
```
(assuming that ```foobar``` is a simple component)

Notinhg is rendered then. How can I force rendering and processing the component?

**Comment by iJackUA** [comment]
I have made it very dirty for myself, some html always present in template
`<modal v-ref="MyModal"></modal>`
And then I can get MyModal by `this.$.MyModal` and trigger some modal show function. Inside modal I can use parent data via  `self.$parent.`.
The main issue I had is `v-ref`.
Is there any way to get component handle without making explicit v-rel ? In any case an Object is created internally in Vue, and it would be good to be able to access it.

The first step for delayed initialization of a component is 
http://vuejs.org/guide/components.html#Async_Components
But as for complete add/remove I don't have ideas right now.

**Comment by PavelPolyakov** [comment]
@yyx990803 
Please, have a look. We're interested in the proper way.

**Comment by iJackUA** [comment]
As a proper way to add component it seems to be something like
```
var a = app.$addChild({ ... component options...});
a.$mount();
```
and if component is already defined, as I understand it should be done with constructor of this component
```
var ModalConstructor = Vue.component('my-modal', { ... component options...});
var a = app.$addChild({},ModalConstructor);   // I am not sure as for empty {} as first argument
a.$mount();
```
but it is not tested, just ideas

**Comment by yyx990803** [comment]
Just add a `v-if`:

``` html
<modal v-if="modalActive"></modal>
```

**Comment by PavelPolyakov** [comment]
@yyx990803 
You mean we should always add the component which might appear in the other component directly to the template, but make it "live" with the flag?

Is there a way to put the component inside the different component on demand?

**Comment by yyx990803** [comment]
Yes, because that makes it more explicit what can possibly appear. You read the template and you know what can be there. Injecting stuff dynamically requires you to go through your JavaScript to figure what could happen.

**Comment by yyx990803** [comment]
Also, if `v-if` evaluates to false in the beginning, it just skips the content without compiling it. It only compiles the content when the condition changes to true.

**Comment by PavelPolyakov** [comment]
@yyx990803 
Thanks for the help, from some point it makes sense.
Would try.

**Comment by PavelPolyakov** [comment]
@yyx990803 

Couple more questions.

Together with gitter chat users, I came up with the next solution:
http://jsbin.com/dejonevoxo/edit?html,js,output

It appends the component container to the main template and fires the component after.

I have three questions regarding this approach:

* How do you find it?
* Why I can't attach the component to the element, like I'm trying to do on the line
```
    el: '.modalContainer', // why this doesn't work $(this.$el).find('.modalContainer') ??
```
While in the documentation it's stated that I can pass the html element there:
http://vuejs.org/api/options.html#el
I receive the next error, when trying to do that:
![image](https://cloud.githubusercontent.com/assets/839290/8630246/a1a7e262-2763-11e5-823b-5a42570e7059.png)
My goal is to be able to attach the component to some non unique element in the html.

* What is the correct way, to remove the component totally. Am I doing it right? Or I just remove the html, while the memory leak is on it's place.

@iJackUA you could be interested in this approach as well

Regards,

**Comment by PavelPolyakov** [comment]
@yyx990803 
Another try is to use the approach with $appendTo:

http://jsbin.com/qotagazidu/1/edit?html,js,output

it doesn't work

![image](https://cloud.githubusercontent.com/assets/839290/8633372/afbacd18-27c4-11e5-8499-47ad9521c095.png)

Why it don't want to append the component?

Regards,

**Comment by PavelPolyakov** [comment]
finally, found the variant which fits me the best so far:
http://jsbin.com/mopotujuye/1/edit?html,js,output

The only thing I want to get rid of is
```javascript
this.modal.$mount($(this.$el).find('.modalContainer').get().pop());
// this.modal.$mount($(this.$el).find('.modalContainer')[0]); looks not very nice
```

The ```.get().pop()``` , how sweet it would be to have an option to use jQuery objects.

**Comment by yyx990803** [comment]
No need more jQuery here... just do

``` js
this.modal.$mount(this.$el.querySelector('.modalContainer'))
```

**Comment by PavelPolyakov** [comment]
@yyx990803 
Thanks

**Comment by abishekrsrikaanth** [comment]
@PavelPolyakov Do you have the last working copy of the jsbin for this issue?

**Comment by PavelPolyakov** [comment]
@abishekrsrikaanth 
This one is actual:
http://jsbin.com/mopotujuye/1/edit?html,js,output

it doesn't work with bootstrap modal, but it does everything which might be helpful.
